Add support for Microsoft-specific modifiers

MathieuNls opened the pull request:
Microsoft-specific modifiers are keywords can be used to modify declarators to form derived types.

The specification for the modifiers themselves is here:
https://docs.microsoft.com/en-us/cpp/cpp/microsoft-specific-modifiers?view=vs-2019.

If you are OK with these changes, I'll go ahead and open a synchronized PR on the cpp grammar to add the Microsoft modifiers there and fix the error that this PR introduces:

```
tree-sitter-cpp# ./tree-sitter-linux-x64 generate

Unresolved conflict for symbol sequence:

  attribute_specifier  •  'extern'  …

Possible interpretations:

  1:  (_declaration_specifiers_repeat1  attribute_specifier)  •  'extern'  …
  2:  (constructor_or_destructor_definition_repeat1  attribute_specifier)  •  'extern'  …

Possible resolutions:

  1:  Specify a higher precedence in `_declaration_specifiers_repeat1` than in the other rules.
  2:  Specify a higher precedence in `constructor_or_destructor_definition_repeat1` than in the other rules.
  3:  Add a conflict for these rules: `_declaration_specifiers`, `constructor_or_destructor_definition`
```

Fixes #41

calixteman commented:
@maxbrunsfeld ping

maxbrunsfeld commented on grammar.js:
```diff
@@ -237,6 +240,60 @@ module.exports = grammar({
       ')'
     ),
 
+    ms_modifiers_declspec: $ => seq(
```
Can you rename this to `ms_declaration_modifier` or `ms_declspec_modifier`?

I think that the singular word "modifier" makes sense because this node just represents *one* modifier, as opposed to a list of many.

Usually we avoid abbreviations like `decl` or `expr`, but I can see the argument that in this case the term "declspec" may be widely used. I'll leave the decision of "declspec_modifier" vs "declaration_modifier" up to you, I just wanted to throw out the suggestion of the non-abbreviated version.

maxbrunsfeld commented on grammar.js:
```diff
@@ -237,6 +240,60 @@ module.exports = grammar({
+      ')',
+    ),
+
+    ms_modifiers_based: $ => seq(
```
Can you call this one `ms_based_modifier`?

maxbrunsfeld commented on grammar.js:
```diff
@@ -237,6 +240,60 @@ module.exports = grammar({
+      $.argument_list,
+    ),
+
+    ms_modifiers_call: $ => choice(
```
Can you rename this one to `ms_call_modifier`?

maxbrunsfeld commented on grammar.js:
```diff
@@ -296,17 +353,32 @@ module.exports = grammar({
+      optional($.ms_modifiers_unaligned_ptr),
+      optional($.ms_modifiers_restrict),
+      optional($.ms_modifiers_unsigned_ptr),
+      optional($.ms_modifiers_signed_ptr),
```
This is another area where I think it's worth "loosening" things a bit. In terms of state machine size, it's easier to match things like `repeat(thing)` than it is to match `optional(thing1), optional(thing2), optional(thing3)...`.

What do you think of replacing these four optional elements with just `repeat($.ms_pointer_modifier)`, where `ms_pointer_modifier` would be a choice between any of the pointer-related keywords above?

MathieuNls replied:
Works for me. I'll amend this.

MathieuNls commented on grammar.js:
```diff
@@ -237,6 +240,60 @@ module.exports = grammar({
+        'selectany',
+        'spectre',
+        'thread',
+        'uuid',
```
Will do.

MathieuNls commented:
@maxbrunsfeld ; I've addressed the comments and opened https://github.com/tree-sitter/tree-sitter-cpp/pull/79 to reflect the changes on the cpp project.

Note that the test on https://github.com/tree-sitter/tree-sitter-cpp/pull/79 won't pass until this is closed as it relies on the 0.17.0 version that I propose here.

MathieuNls commented:
@maxbrunsfeld ping

maxbrunsfeld commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "tree-sitter-c",
-  "version": "0.16.0",
+  "version": "0.17.0",
```
I'd actually like to avoid bumping the minor version number for this. At this point in Tree-sitter's development, I've been mainly using the semver version numbers to communicate binary compatibility between the Tree-sitter library and the parsers. So right now, most of the parsers are at version `0.16.x`.

Would you mind switching this to `0.16.1`?

MathieuNls replied:
Done. Also adjusted on the cpp side.

maxbrunsfeld commented:
Thanks so much; great work on this. I will try to look at the C++ PR later today.

MathieuNls commented:
The last thing required for  tree-sitter/tree-sitter-cpp#79 tests to pass is to tag 88592f162e218e3d0de28f1e5d1835e6ef2c57b6 as `0.16.1`
